dm/regexprrouter/router.go

```diff
@@ -0,0 +1,278 @@
+	}
+	for _, rule := range rules {
+		if err := r.AddRule(rule); err != nil {
+			return nil, err
+		}
+	}
+	return r, nil
+}
+
+func (r *RegExprTable) AddRule(rule *TableRule) error {
+	err := rule.Valid()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !r.caseSensitive {
+		rule.ToLower()
+	}
+	newFilter := &FilterWrapper{
+		rawRule: rule,
+	}
+	newFilter.Target = Table{
+		Schema: rule.TargetSchema,
+		Name:   rule.TargetTable,
+	}
+	if len(rule.TablePattern) == 0 {
+		// raw schema rule
+		newFilter.Type = SchmFilter
+		rawFilter, err := filter.New(r.caseSensitive, &FilterRule{
+			DoDBs: []string{rule.SchemaPattern},
+		})
+		if err != nil {
+			return errors.Annotatef(err, "add rule %+v into table router", rule)
+		}
+		newFilter.Filter = rawFilter
+	} else {
+		newFilter.Type = TblFilter
+		rawFilter, err := filter.New(r.caseSensitive, &FilterRule{
+			DoTables: []*Table{
+				{
+					Schema: rule.SchemaPattern,
+					Name:   rule.TablePattern,
+				},
+			},
+			DoDBs: []string{rule.SchemaPattern},
+		})
+		if err != nil {
+			return errors.Annotatef(err, "add rule %+v into table router", rule)
+		}
+		newFilter.Filter = rawFilter
+	}
+	r.filters = append(r.filters, newFilter)
+	return nil
+}
+
+func (r *RegExprTable) Route(schema, table string) (string, string, error) {
+	schemaL, tableL := schema, table
+	if !r.caseSensitive {
+		schemaL, tableL = strings.ToLower(schema), strings.ToLower(table)
+	}
+	curTable := &Table{
+		Schema: schemaL,
+		Name:   tableL,
+	}
+	tblRules := make([]*FilterWrapper, 0)
+	schmRules := make([]*FilterWrapper, 0)
+	for _, filterWrapper := range r.filters {
+		if filterWrapper.Filter.Match(curTable) {
+			if filterWrapper.Type == TblFilter {
+				tblRules = append(tblRules, filterWrapper)
+			} else {
+				schmRules = append(schmRules, filterWrapper)
+			}
+		}
+	}
+	var (
+		targetSchema string
+		targetTable  string
+	)
+	if len(table) == 0 || len(tblRules) == 0 {
+		// 1. no need to match table or
+		// 2. match no table
+		if len(schmRules) > 1 {
+			return "", "", errors.NotSupportedf("`%s`.`%s` matches %d schema route rules which is more than one.\n", schema, table, len(schmRules))
+		}
+		if len(schmRules) == 1 {
+			targetSchema, targetTable = schmRules[0].Target.Schema, schmRules[0].Target.Name
+		}
+	} else {
+		if len(tblRules) > 1 {
+			return "", "", errors.NotSupportedf("`%s`.`%s` matches %d table route rules which is more than one.\n", schema, table, len(tblRules))
+		}
+		targetSchema, targetTable = tblRules[0].Target.Schema, tblRules[0].Target.Name
+	}
+	if len(targetSchema) == 0 {
+		targetSchema = schema
+	}
+	if len(targetTable) == 0 {
+		targetTable = table
+	}
+	return targetSchema, targetTable, nil
+}
+
+func (r *RegExprTable) AllRules() (map[string][]interface{}, map[string]map[string][]interface{}) {
+	schemaRouteRules := make(map[string][]interface{})
+	tableRouteRules := make(map[string]map[string][]interface{})
+	for _, filter := range r.filters {
+		curRule := displayRule{
```

Comment:
seems this is only used for debug? You can use json dump for `filter.rawRule`, no need to add a new structure

Reply:
Or can I just return two arrays representing the table rules and the schema rules respectively?